### DISCUSSION TITLE
[5.0]: Better content management with matrix fields (UI/UX for editor + admin)

### DISCUSSION THREAD

**Discussion by alexanderbuergin** [craftcms/cms#14179] [Ideas] [answered]
I recommend that each block type should have the capability to be organized hierarchically within the matrix field. This structural approach would enable content architects to define a clear content hierarchy, improving the end-user experience and content discoverability.

---

**Grouped Entry Types**

I suggest that the interface should allow content creators to sort entry/block types into groups. These groups would then be respected in the output when the builder is deployed, providing content editors with a clearer overview and improved organization.

<img width="1714" alt="Bildschirmfoto 2024-01-23 um 22 26 54" src="https://github.com/craftcms/cms/assets/85960512/2635cad4-7428-4684-b7ed-eededba6d135">

Displaying Blocks in Groups within Edit Mode

<img width="1358" alt="Bildschirmfoto 2024-01-23 um 22 22 55" src="https://github.com/craftcms/cms/assets/85960512/0e4e4fc9-c519-4744-abee-30da99225cdf">

<img width="1352" alt="Bildschirmfoto 2024-01-23 um 22 42 10" src="https://github.com/craftcms/cms/assets/85960512/ffa13155-7c3a-4906-93ee-58eedd8442cb">

---

**Nested Blocks**

With the advent of Craft CMS 5.0, nested blocks are anticipated. I propose a feature that would maintain and place multi-level structures, allowing nested blocks to be organized well and creator friendly. This would facilitate a more efficient content creation process and offer a modular approach to content management.

<img width="1707" alt="Bildschirmfoto 2024-01-23 um 22 20 05" src="https://github.com/craftcms/cms/assets/85960512/ffa53788-3ad5-4b8c-8ad3-617512a82af6">

---

**Reusable Entry Blocks in Matrix Fields**

It should be possible to use entry blocks multiple times within a matrix field. For instance, one might want to use a type like 'Title' across several blocks as a child element. This flexibility would significantly enhance the content structuring capabilities of Craft CMS.

<img width="1671" alt="Bildschirmfoto 2024-01-23 um 22 30 09" src="https://github.com/craftcms/cms/assets/85960512/b062708c-1339-4af8-90b1-9b3e89023f79"> 

---

Change `Site Settings` to something like `Matrix Settings`

**Comment by alexanderbuergin** [comment]
An alternative method of organizing the block and entry types could be to revise the user interface of the current structure view of the entries. This UI could then be adapted for editing the entry and block types in the matrix field. This would make it possible to transfer the familiar and intuitive structure view to the matrix fields in order to provide a consistent and user-friendly experience when organizing the content.

---

**UI basis**
<img width="1501" alt="Bildschirmfoto 2024-01-23 um 23 15 21" src="https://github.com/craftcms/cms/assets/85960512/45226ef7-5c75-4e4e-a79b-a557f568cced">

**UI customized**
<img width="1478" alt="Bildschirmfoto 2024-01-23 um 23 15 08" src="https://github.com/craftcms/cms/assets/85960512/4b314beb-04b6-408d-8493-6ebeeccb3973">

**Comment by alexanderloewe** [comment]
+1

**Comment by davidhellmann** [comment]
The latest update is a huge step for the matrix field (alpha 9). I like the entry type sorting, the icons, and the colors. However, we need a grouping for entry types within matrix fields. It would be great to have something like that within the core and not in a third-party plugin. 

The nested thing here I don't understand. When I'm right you can achieve that by using another Matrix Field within a Matrix Field Entry Type. Not sure if that was the goal.

**Comment by mmikkel** [comment]
+1 for the ability to group entry types per Matrix field. At this point, this is pretty much the only worthwhile feature in plugins like MatrixMate and Spoon for Craft 3 and 4, that isn't already a core feature in Craft 5.

As far as implementation goes; ideally I'd want to see a discrete disclosure menu per group, but even just having the groups as optgroup-ish headings inside the single "Add an entry..." disclosure menu would be very helpful to content authors struggling in huge Mega-Matrixes 😅 In fact, maybe the feature just needs to be "Headings", ala element indexes?

Both MatrixMate and Spoon have been quite popular plugins for years, and as the author of MatrixMate I'm pretty sure that block type groups is probably its most popular feature. Would be amazing to have that in core (it would also save me from having to port a plugin that is otherwise redundant in 5.0 😁)

**Reply by brandonkelly** [reply]
Adding headings within the single disclosure menu is most likely how I’d want to go about it, similar to how site selectors/menus work when there are multiple site groups.

Do you think the groups should be defined on a per-field basis, or just centrally via a new global “Entry Type Groups” feature similar to site groups?

**Reply by alexanderloewe** [reply]
I think that the groups should not be global, but that they should be entered and edited within the matrix field.

**Reply by mmikkel** [reply]
> Do you think the groups should be defined on a per-field basis, or just centrally via a new global “Entry Type Groups” feature similar to site groups?

Oof, tough call. 

FWIW, MatrixMate has essentially provided both – i.e. it's possible to define a "global" group config, but also field-specific configs (that override the global config, if there is one). I'm not saying Craft should necessarily do the same thing, partly because I realise everything Gets More Complex once there's UI involved (MatrixMate is completely configured in a PHP config file).

I guess if I had to choose one or the other, my gut feeling says that it should most likely be a per-field feature (maybe even per-section? No reason why the "Entry type" select in entry edit pages couldn't have groups/headings as well). 

While global entry type groups would likely be simpler to implement and deal with, and certainly nice in certain situations (such as entry types being automatically grouped when re-used in new/different Matrix fields), this really feels like a contextual feature to me. 

Some fields will have lots of entry types and would benefit a lot from grouping. Other fields might just have 2 or 3 types, and having them "grouped" will just look weird and redundant. 

Depending on the context, an "Image" entry type might make sense in an "Images" group in one Matrix field, but should ideally appear in a "Visuals" or "Media" group in a different field. Etc. 🙂

**Reply by amici-infotech** [reply]
Do we have any thought on adding headings or making grouped blocks options? I am really looking forward to this feature.

**Reply by alexanderloewe** [reply]
Me too @brandonkelly

**Reply by leevigraham** [reply]
> Do you think the groups should be defined on a per-field basis,

Per field.

**Comment by alexanderloewe** [comment]
The way I see it, there won't be a direct Matrix in Matrix, like Neo offers, right @brandonkelly? 
Instead, you have to pack several matrix fields together?

If that is the case: This has the disadvantage that you may have to build a lot of matrix fields this way if you have to have 10 different types of child blocks, for example. 

Or if, for example, not all child blocks are used in the same combination. In my opinion, this is a major disadvantage of this approach.

**Reply by brandonkelly** [reply]
There is full-blown nested Matrix support in Craft 5.

If the list of available entry types needs to change depending on the context, for now you have two options:

1. Create multiple Matrix fields. (Maybe not ideal, but entry types are reusable now so you’re only having to reselect entry types; not recreate them for each field.)
2. Customize the available entry types programmatically using the `craft\fields\Matrix::EVENT_DEFINE_ENTRY_TYPES` event.

**Reply by alexanderloewe** [reply]
What I don't understand in this context:

In my opinion it would be much better to create, for example, an EntryType `Title`, `Text`, `List` and I can then set this EntryType as a child for all other EntryTypes in the matrix field itself. 

This way you would have 100 Entry Types, but you could assemble them as flexibly as the content requires, all in just one Matrix field.

In my opinion, this type-based approach would be extremely content-friendly

**Reply by alexanderloewe** [reply]
One example:
```
Block Teaser (EntryType)
–– Child Title (EntryType)
–– Child Text (EntryType)

Block CTA (EntryType)
–– Child Title (EntryType)
–– Child Text (EntryType)
–– Child List (EntryType)
```

**Reply by alexanderloewe** [reply]
In the same style as above

<img width="1671" alt="299108138-b062708c-1339-4af8-90b1-9b3e89023f79" src="https://github.com/craftcms/cms/assets/69263307/9e297389-58de-41dd-9426-2b2460eaa8ca">

**Reply by brandonkelly** [reply]
Matrix blocks don’t have hierarchy. I know Neo does that, but we have no interest in bringing it to Matrix. It comes with a significant complexity and performance hit.

**Reply by alexanderloewe** [reply]
Then the concept of flexibility is abandoned and the process for multilevel fields becomes more difficult.

I think you should reconsider this. In my eyes, Craft is a CMS solution that favours flexibility and abandoning this does not seem to me to be the solution. 

Since content modelling sometimes requires nesting, I believe that flexibility should also be taken into account here.

The chosen solution of simply nesting a matrix field in a matrix field is more or less the solution that has often been implemented with Super Table. Which has often led to problems due to its rigidity.

**Reply by brandonkelly** [reply]
There’s definitely no compromise on _flexibility_, if a Matrix block can have multiple _sets_ of nested Matrix blocks. And having supported several Craft sites with heavy use of hierarchical Neo fields over the years, the idea that hierarchical blocks is a better UX is controversial at best. Those forms can very quickly become unruly, confusing, and frustrating. Moving nested content into slideouts has a huge potential to improve the overal UX.

**Reply by thupsi** [reply]
I personally agree with Brandon, that there is no loss in flexibility with the current approach. Perhaps I'm misunderstanding something, but to me it seems what you are proposing can already be achieved, with different steps. 

So, you would have a matrix field 'Blocks' configured to accept 2 entry types: 'Teaser' and 'CTA'. And in those entry types you would have the fields you are describing. 

Also, I don't see a major problem with building a lot of matrix fields/entry types. You would need to build the blocks anyway, which is a similar amount of work, plus entry types are global and reusable. Again, I'm not being dismissive, just coming at this from a different angle.

**Reply by leevigraham** [reply]
@alexanderloewe When ever I've needed nested blocks they seem to be very specific to the parent. In this case I use an new matrix field with fewer block options

Accordions:

![image](https://github.com/user-attachments/assets/b6c683a9-fced-4aa6-ae13-832eb46407b2)

Logo Grid:

![image](https://github.com/user-attachments/assets/d7388871-09e4-4943-8d82-538307f9d9aa)

**Comment by anita-chouhan** [comment]
+1 for the ability to group entry types per Matrix field.

**Comment by sandrodunkel** [comment]
Any updates on this? 👀 For simple grouping there's always the [Matrix Extended](https://github.com/vandres/craft-matrix-extended) plugin, but this very much feels like it should be core functionality.

**Comment by brandonkelly** [comment, marked as answer]
Craft 5.8.0 is out with the ability to organize Matrix fields’ entry types into groups. (#17425)
